sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/PostgresConnectionProvider.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.jdbc.connection
+
+import java.sql.{Connection, Driver}
+import javax.security.auth.login.{AppConfigurationEntry, Configuration}
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions
+import org.apache.spark.sql.execution.datasources.jdbc.connection.PostgresConnectionProvider.PGJDBCConfiguration
+import org.apache.spark.util.SecurityUtils
+
+private[jdbc] class PostgresConnectionProvider(driver: Driver, options: JDBCOptions)
+    extends BasicConnectionProvider(driver, options) {
+  def setAuthenticationConfigIfNeeded(): Unit = {
+    val parent = Configuration.getConfiguration
+    val configEntry = parent.getAppConfigurationEntry(PostgresConnectionProvider.appEntry)
+    if (configEntry == null || configEntry.isEmpty) {
+      val config = new PGJDBCConfiguration(parent, options.keytab, options.principal)
+      Configuration.setConfiguration(config)
+    }
+  }
+
+  override def getConnection(): Connection = {
+    setAuthenticationConfigIfNeeded()
+    super.getConnection()
+  }
+}
+
+private[sql] object PostgresConnectionProvider {
+  class PGJDBCConfiguration(
+      parent: Configuration,
+      keytab: String, principal: String) extends Configuration {
+    private val entry =
+      new AppConfigurationEntry(
+        SecurityUtils.getKrb5LoginModuleName(),
+        AppConfigurationEntry.LoginModuleControlFlag.REQUIRED,
+        Map[String, Object](
+          "useTicketCache" -> "false",
+          "useKeyTab" -> "true",
+          "keyTab" -> keytab,
+          "principal" -> principal,
+          "debug" -> "true"
+        ).asJava
+      )
+
+    override def getAppConfigurationEntry(name: String): Array[AppConfigurationEntry] = {
+      if (name.equals(appEntry)) {
+        Array(entry)
+      } else {
+        parent.getAppConfigurationEntry(name)
+      }
+    }
+  }
+
+  val driverClass = "org.postgresql.Driver"
+  val appEntry = "pgjdbc"
```

Comment:
Following a previous comment of yours... if you have a Spark app that needs to connect to 2 different pgsql data sources, each using different credentials, will you have a problem here?

Reply:
That's a really good point which the actual code doesn't cover. Postgres supports to configure `jaasApplicationName` which must be used to overcome this issue. Adding the support...

Reply:
In order to support this jdbc `url` must be parsed which can be done several ways:
* Using `postgres` by calling the appropriate API function through reflection => No ugly dependency
* Using `postgres` by calling the appropriate API function through provided dependency => No ugly reflection
* Re-implement it => It's an overkill

I've chose the first approach but if you think the second is better we can change it.